ZupIT/beagle-web-core
# Decouples pre-fetch operations from the cache sub-system

**- What I did**
closes #295 

Today, the pre-fetch is done via the cache sub-system, i.e. when a navigation with pre-fetch is found, Beagle fetches the view from the remote server and caches it in the local storage. When the view is needed, since it's cached already, it will be available.

As explained in the issue #295, this is not right because if the developer disables the cache, we should also disable the pre-fetch. Right now, if the developer sets the strategy to `network-only` and uses a pre-fetch, we ignore the configuration and cache the pre-fetched view anyway.

As suggested in #295, I decoupled these systems. Now there's a pre-fetch service (PreFetcher) and it has nothing to do with caching. Even if the cache is disabled, it will work and it will never store anything locally, i.e. it's safe to use in any situation.

I also fixed the following issues with the pre-fetch feature:
- Pre-fetch requests wouldn't send the default headers.
- Pre-fetch requests would ignore the cache strategy set in the config.
- When the navigation happened before the pre-fetch finished, it would trigger another request. Instead, it should wait the current request to finish.
- Suppose the following scenario: 1. The view X is pre-fetched successfully, but it's not used. 5 minutes later, another view (which is not X) also pre-fetches X, but this time the pre-fetch fails. When navigating to X, since the pre-fetch failed, it should try to get the view from the server, but instead, it's going to use the first version of X that was pre-fetched, but not used. Since pre-fetch is not a cache and has no "expiration" mechanism, this should not happen.

**- How I did it**
I created a new service: PreFetcher. `preFetcher.fetch` is called when a navigation with `{ shouldPrefetch: true }` is found and `preFetcher.recover` is called when the navigation actually happens. All views are stored in memory and never in the local storage.

**- How to verify it**
I created new tests and altered existing tests to reflect this change. Run them.

I also ran the playground with the following modifications:
- I logged a message every time a view was fetched;
- I delayed any request to a view by 2 seconds so I could see the difference between using and not using the pre-fetch.

To see the execution of the playground with the modification above, please watch [this video](https://drive.google.com/file/d/1tifhKdKZ7iOAjJe5mjM-KR5kWV6gLXBL/view?usp=sharing).

**- Description for the changelog**
- Modified how the pre-fetch works so it never uses the local storage.
- Fixed a problem where pre-fetch requests wouldn't include the default headers.
- Fixed a problem where pre-fetch requests wouldn't respect the strategy set in the configuration.
- Fixed a problem where recovering a pre-fetched view would not wait for its corresponding fetch, triggering a new request
- Fixed a problem where old pre-fetched views could be used as the value of new pre-fetch operations.
